File: app/src/main/java/com/example/msdc/api/ApiService.java
package com.example.msdc.api;

import retrofit2.Call;
import retrofit2.http.GET;
import retrofit2.http.Path;
import retrofit2.http.Query;

public interface ApiService {
    @GET("movie/popular")
    Call<MovieResponse> getPopularMovies(@Query("api_key") String apiKey, @Query("language") String language,
                                         @Query("page") int page, @Query("limit") int limit);

    @GET("movie/now_playing")
    Call<MovieResponse> getNowPlayingMovies(@Query("api_key") String apiKey, @Query("language") String language,
                                            @Query("page") int page, @Query("limit") int limit);

    @GET("movie/upcoming")
    Call<MovieResponse> getUpcomingMovies(@Query("api_key") String apiKey, @Query("language") String language,
                                          @Query("page") int page, @Query("limit") int limit);

    @GET("movie/top_rated")
    Call<MovieResponse> getTopRatedMovies(@Query("api_key") String apiKey, @Query("language") String language,
                                          @Query("page") int page, @Query("limit") int limit);

    @GET("trending/{media_type}/{time_window}")
    Call<MovieResponse> getTrendingMovies(@Path("media_type") String mediaType,
                                          @Path("time_window") String timeWindow,
                                          @Query("api_key") String apiKey);

    @GET("movie/{movie_id}")
    Call<MovieDetails> getMovieDetails(@Path("movie_id") int id, @Query("api_key") String apiKey);

    @GET("movie/{movie_id}/recommendations")
    Call<MovieResponse> getMovieRecommendations(@Path("movie_id") int id, @Query("api_key") String apiKey);

    @GET("movie/{movie_id}/similar")
    Call<MovieResponse> getMovieSimilar(@Path("movie_id") int id, @Query("api_key") String apiKey);

    @GET("movie/{movie_id}/images")
    Call<ImageResponse> getMovieImages(@Path("movie_id") int id, @Query("api_key") String apiKey);

    @GET("movie/{movie_id}/credits")
    Call<CreditResponse> getMovieCredit(@Path("movie_id") int id, @Query("api_key") String apiKey);

    @GET("movie/{movie_id}/keywords")
    Call<KeywordResponse> getMovieKeywords(@Path("movie_id") int id, @Query("api_key") String apiKey);

    @GET("search/movie")
    Call<MovieResponse> searchMovie(@Query("api_key") String apiKey, @Query("language") String language,
                                    @Query("query") String query, @Query("page") int page,
                                    @Query("limit") int limit);

    @GET("discover/movie")
    Call<MovieResponse> filterMovie(@Query("api_key") String apiKey, @Query("language") String language,
                                      @Query("page") int page, @Query("limit") int limit,
                                      @Query("with_genres") int genreId, @Query("year") int year,
                                      @Query("region") String region, @Query("sort_by") String sortBy);

    @GET("discover/movie")
    Call<MovieResponse> filterMovieGenre(@Query("api_key") String apiKey, @Query("language") String language,
                                    @Query("page") int page, @Query("limit") int limit,
                                    @Query("with_genres") int genreId, @Query("sort_by") String sortBy);

    @GET("discover/movie")
    Call<MovieResponse> filterMovieKeyword(@Query("api_key") String apiKey, @Query("language") String language,
                                         @Query("page") int page, @Query("limit") int limit,
                                         @Query("with_keywords") int keywordId, @Query("sort_by") String sortBy);

    @GET("discover/movie")
    Call<MovieResponse> filterMovieYear(@Query("api_key") String apiKey, @Query("language") String language,
                                         @Query("page") int page, @Query("limit") int limit,
                                         @Query("year") int year, @Query("sort_by") String sortBy);

    @GET("discover/movie")
    Call<MovieResponse> filterMovieRegion(@Query("api_key") String apiKey, @Query("language") String language,
                                          @Query("page") int page, @Query("limit") int limit,
                                          @Query("region") String region, @Query("sort_by") String sortBy);

    @GET("discover/movie")
    Call<MovieResponse> filterMovieGenreAndYear(@Query("api_key") String apiKey, @Query("language") String language,
                                         @Query("page") int page, @Query("limit") int limit,
                                         @Query("with_genres") int genreId, @Query("year") int year,
                                                @Query("sort_by") String sortBy);

    @GET("discover/movie")
    Call<MovieResponse> filterMovieGenreAndRegion(@Query("api_key") String apiKey, @Query("language") String language,
                                                @Query("page") int page, @Query("limit") int limit,
                                                @Query("with_genres") int genreId, @Query("region") String region,
                                                  @Query("sort_by") String sortBy);

    @GET("discover/movie")
    Call<MovieResponse> filterMovieYearAndRegion(@Query("api_key") String apiKey, @Query("language") String language,
                                                 @Query("page") int page, @Query("limit") int limit,
                                                 @Query("year") int year, @Query("region") String region,
                                                 @Query("sort_by") String sortBy);
    @GET("tv/airing_today")
    Call<TVResponse> getTvAiringToday(@Query("api_key") String apiKey, @Query("language") String language,
                                      @Query("page") int page,  @Query("limit") int limit);

    @GET("tv/popular")
    Call<TVResponse> getTvPopular(@Query("api_key") String apiKey, @Query("language") String language,
                                  @Query("page") int page, @Query("limit") int limit);

    @GET("tv/top_rated")
    Call<TVResponse> getTvTopRated(@Query("api_key") String apiKey, @Query("language") String language,
                                   @Query("page") int page,  @Query("limit") int limit);

    @GET("tv/on_the_air")
    Call<TVResponse> getTvOnAir(@Query("api_key") String apiKey, @Query("language") String language,
                                @Query("page") int page,  @Query("limit") int limit);

    @GET("trending/{media_type}/{time_window}")
    Call<TVResponse> getTrendingTV(@Path("media_type") String mediaType,
                                   @Path("time_window") String timeWindow,
                                   @Query("api_key") String apiKey);

    @GET("tv/{series_id}")
    Call<TVDetails> getTvDetails(@Path("series_id") int id, @Query("api_key") String apiKey);

    @GET("tv/{series_id}/images")
    Call<ImageResponse> getTvImages(@Path("series_id") int id, @Query("api_key") String apiKey);

    @GET("tv/{series_id}/keywords")
    Call<KeywordResponse> getTvKeywords(@Path("series_id") int id, @Query("api_key") String apiKey);

    @GET("tv/{series_id}/credits")
    Call<CreditResponse> getTvCredit(@Path("series_id") int id, @Query("api_key") String apiKey);

    @GET("tv/{series_id}/season/{season_number}")
    Call<SeasonResponse> getTvSeasonAndEpisode(@Path("series_id") int id, @Path("season_number")
                                                int seasonNumber, @Query("api_key") String apiKey);

    @GET("search/tv")
    Call<TVResponse> searchTv(@Query("api_key") String apiKey, @Query("language") String language,
                              @Query("query") String query, @Query("page") int page,
                              @Query("limit") int limit);

    @GET("discover/tv")
    Call<TVResponse> filterTv(@Query("api_key") String apiKey, @Query("language") String language,
                              @Query("page") int page, @Query("limit") int limit,
                              @Query("with_genres") int genreId, @Query("year") int year,
                              @Query("region") String region, @Query("sort_by") String sortBy);

    @GET("discover/tv")
    Call<TVResponse> filterTvGenre(@Query("api_key") String apiKey, @Query("language") String language,
                              @Query("page") int page, @Query("limit") int limit,
                              @Query("with_genres") int genreId, @Query("sort_by") String sortBy);

    @GET("discover/tv")
    Call<TVResponse> filterTvKeyword(@Query("api_key") String apiKey, @Query("language") String language,
                                   @Query("page") int page, @Query("limit") int limit,
                                   @Query("with_keywords") int keywordId, @Query("sort_by") String sortBy);

    @GET("discover/tv")
    Call<TVResponse> filterTvYear(@Query("api_key") String apiKey, @Query("language") String language,
                                   @Query("page") int page, @Query("limit") int limit,
                                  @Query("year") int year, @Query("sort_by") String sortBy);

    @GET("discover/tv")
    Call<TVResponse> filterTvRegion(@Query("api_key") String apiKey, @Query("language") String language,
                                   @Query("page") int page, @Query("limit") int limit,
                                    @Query("region") String region, @Query("sort_by") String sortBy);

    @GET("discover/tv")
    Call<TVResponse> filterTvGenreAndYear(@Query("api_key") String apiKey, @Query("language") String language,
                                   @Query("page") int page, @Query("limit") int limit,
                                   @Query("with_genres") int genreId, @Query("year") int year,
                                          @Query("sort_by") String sortBy);

    @GET("discover/tv")
    Call<TVResponse> filterTvGenreAndRegion(@Query("api_key") String apiKey, @Query("language") String language,
                                          @Query("page") int page, @Query("limit") int limit,
                                          @Query("with_genres") int genreId, @Query("region") String region,
                                          @Query("sort_by") String sortBy);

    @GET("discover/tv")
    Call<TVResponse> filterTvYearAndRegion(@Query("api_key") String apiKey, @Query("language") String language,
                                          @Query("page") int page, @Query("limit") int limit,
                                           @Query("year") int year, @Query("region") String region,
                                          @Query("sort_by") String sortBy);

    @GET("trending/{media_type}/{time_window}")
    Call<PersonResponse> getTrendingPerson(@Path("media_type") String mediaType,
                                           @Path("time_window") String timeWindow,
                                           @Query("api_key") String apiKey);

    @GET("tv/{series_id}/recommendations")
    Call<TVResponse> getTVRecommendations(@Path("series_id") int id, @Query("api_key") String apiKey);

    @GET("tv/{series_id}/similar")
    Call<TVResponse> getTVSimilar(@Path("series_id") int id, @Query("api_key") String apiKey);
}

File: app/src/main/java/com/example/msdc/api/ApiService.kt
package com.example.msdc.api

import retrofit2.Call
import retrofit2.http.GET
import retrofit2.http.Path
import retrofit2.http.Query

interface ApiService {
    @GET("movie/popular")
    fun getPopularMovies(
        @Query("api_key") apiKey: String?, @Query("language") language: String?,
        @Query("page") page: Int, @Query("limit") limit: Int
    ): Call<MovieResponse?>?

    @GET("movie/now_playing")
    fun getNowPlayingMovies(
        @Query("api_key") apiKey: String?, @Query("language") language: String?,
        @Query("page") page: Int, @Query("limit") limit: Int
    ): Call<MovieResponse?>?

    @GET("movie/upcoming")
    fun getUpcomingMovies(
        @Query("api_key") apiKey: String?, @Query("language") language: String?,
        @Query("page") page: Int, @Query("limit") limit: Int
    ): Call<MovieResponse?>?

    @GET("movie/top_rated")
    fun getTopRatedMovies(
        @Query("api_key") apiKey: String?, @Query("language") language: String?,
        @Query("page") page: Int, @Query("limit") limit: Int
    ): Call<MovieResponse?>?

    @GET("trending/{media_type}/{time_window}")
    fun getTrendingMovies(
        @Path("media_type") mediaType: String?,
        @Path("time_window") timeWindow: String?,
        @Query("api_key") apiKey: String?
    ): Call<MovieResponse?>?

    @GET("movie/{movie_id}")
    fun getMovieDetails(
        @Path("movie_id") id: Int,
        @Query("api_key") apiKey: String?
    ): Call<MovieDetails?>?

    @GET("movie/{movie_id}/recommendations")
    fun getMovieRecommendations(
        @Path("movie_id") id: Int,
        @Query("api_key") apiKey: String?
    ): Call<MovieResponse?>?

    @GET("movie/{movie_id}/similar")
    fun getMovieSimilar(
        @Path("movie_id") id: Int,
        @Query("api_key") apiKey: String?
    ): Call<MovieResponse?>?

    @GET("movie/{movie_id}/images")
    fun getMovieImages(
        @Path("movie_id") id: Int,
        @Query("api_key") apiKey: String?
    ): Call<ImageResponse?>?

    @GET("movie/{movie_id}/credits")
    fun getMovieCredit(
        @Path("movie_id") id: Int,
        @Query("api_key") apiKey: String?
    ): Call<CreditResponse?>?

    @GET("movie/{movie_id}/keywords")
    fun getMovieKeywords(
        @Path("movie_id") id: Int,
        @Query("api_key") apiKey: String?
    ): Call<KeywordResponse?>?

    @GET("search/movie")
    fun searchMovie(
        @Query("api_key") apiKey: String?, @Query("language") language: String?,
        @Query("query") query: String?, @Query("page") page: Int,
        @Query("limit") limit: Int
    ): Call<MovieResponse?>?

    @GET("discover/movie")
    fun filterMovie(
        @Query("api_key") apiKey: String?, @Query("language") language: String?,
        @Query("page") page: Int, @Query("limit") limit: Int,
        @Query("with_genres") genreId: Int, @Query("year") year: Int,
        @Query("region") region: String?, @Query("sort_by") sortBy: String?
    ): Call<MovieResponse?>?

    @GET("discover/movie")
    fun filterMovieGenre(
        @Query("api_key") apiKey: String?, @Query("language") language: String?,
        @Query("page") page: Int, @Query("limit") limit: Int,
        @Query("with_genres") genreId: Int, @Query("sort_by") sortBy: String?
    ): Call<MovieResponse?>?

    @GET("discover/movie")
    fun filterMovieKeyword(
        @Query("api_key") apiKey: String?, @Query("language") language: String?,
        @Query("page") page: Int, @Query("limit") limit: Int,
        @Query("with_keywords") keywordId: Int, @Query("sort_by") sortBy: String?
    ): Call<MovieResponse?>?

    @GET("discover/movie")
    fun filterMovieYear(
        @Query("api_key") apiKey: String?, @Query("language") language: String?,
        @Query("page") page: Int, @Query("limit") limit: Int,
        @Query("year") year: Int, @Query("sort_by") sortBy: String?
    ): Call<MovieResponse?>?

    @GET("discover/movie")
    fun filterMovieRegion(
        @Query("api_key") apiKey: String?, @Query("language") language: String?,
        @Query("page") page: Int, @Query("limit") limit: Int,
        @Query("region") region: String?, @Query("sort_by") sortBy: String?
    ): Call<MovieResponse?>?

    @GET("discover/movie")
    fun filterMovieGenreAndYear(
        @Query("api_key") apiKey: String?, @Query("language") language: String?,
        @Query("page") page: Int, @Query("limit") limit: Int,
        @Query("with_genres") genreId: Int, @Query("year") year: Int,
        @Query("sort_by") sortBy: String?
    ): Call<MovieResponse?>?

    @GET("discover/movie")
    fun filterMovieGenreAndRegion(
        @Query("api_key") apiKey: String?, @Query("language") language: String?,
        @Query("page") page: Int, @Query("limit") limit: Int,
        @Query("with_genres") genreId: Int, @Query("region") region: String?,
        @Query("sort_by") sortBy: String?
    ): Call<MovieResponse?>?

    @GET("discover/movie")
    fun filterMovieYearAndRegion(
        @Query("api_key") apiKey: String?, @Query("language") language: String?,
        @Query("page") page: Int, @Query("limit") limit: Int,
        @Query("year") year: Int, @Query("region") region: String?,
        @Query("sort_by") sortBy: String?
    ): Call<MovieResponse?>?

    @GET("tv/airing_today")
    fun getTvAiringToday(
        @Query("api_key") apiKey: String?, @Query("language") language: String?,
        @Query("page") page: Int, @Query("limit") limit: Int
    ): Call<TVResponse?>?

    @GET("tv/popular")
    fun getTvPopular(
        @Query("api_key") apiKey: String?, @Query("language") language: String?,
        @Query("page") page: Int, @Query("limit") limit: Int
    ): Call<TVResponse?>?

    @GET("tv/top_rated")
    fun getTvTopRated(
        @Query("api_key") apiKey: String?, @Query("language") language: String?,
        @Query("page") page: Int, @Query("limit") limit: Int
    ): Call<TVResponse?>?

    @GET("tv/on_the_air")
    fun getTvOnAir(
        @Query("api_key") apiKey: String?, @Query("language") language: String?,
        @Query("page") page: Int, @Query("limit") limit: Int
    ): Call<TVResponse?>?

    @GET("trending/{media_type}/{time_window}")
    fun getTrendingTV(
        @Path("media_type") mediaType: String?,
        @Path("time_window") timeWindow: String?,
        @Query("api_key") apiKey: String?
    ): Call<TVResponse?>?

    @GET("tv/{series_id}")
    fun getTvDetails(
        @Path("series_id") id: Int,
        @Query("api_key") apiKey: String?
    ): Call<TVDetails?>?

    @GET("tv/{series_id}/images")
    fun getTvImages(
        @Path("series_id") id: Int,
        @Query("api_key") apiKey: String?
    ): Call<ImageResponse?>?

    @GET("tv/{series_id}/credits")
    fun getTvCredit(
        @Path("series_id") id: Int,
        @Query("api_key") apiKey: String?
    ): Call<CreditResponse?>?

    @GET("tv/{series_id}/season/{season_number}")
    fun getTvSeasonAndEpisode(
        @Path("series_id") id: Int,
        @Path("season_number") seasonNumber: Int,
        @Query("api_key") apiKey: String?
    ): Call<SeasonResponse?>?

    @GET("search/tv")
    fun searchTv(
        @Query("api_key") apiKey: String?, @Query("language") language: String?,
        @Query("query") query: String?, @Query("page") page: Int,
        @Query("limit") limit: Int
    ): Call<TVResponse?>?

    @GET("discover/tv")
    fun filterTv(
        @Query("api_key") apiKey: String?, @Query("language") language: String?,
        @Query("page") page: Int, @Query("limit") limit: Int,
        @Query("with_genres") genreId: Int, @Query("year") year: Int,
        @Query("region") region: String?, @Query("sort_by") sortBy: String?
    ): Call<TVResponse?>?

    @GET("discover/tv")
    fun filterTvGenre(
        @Query("api_key") apiKey: String?, @Query("language") language: String?,
        @Query("page") page: Int, @Query("limit") limit: Int,
        @Query("with_genres") genreId: Int, @Query("sort_by") sortBy: String?
    ): Call<TVResponse?>?

    @GET("discover/tv")
    fun filterTvKeyword(
        @Query("api_key") apiKey: String?, @Query("language") language: String?,
        @Query("page") page: Int, @Query("limit") limit: Int,
        @Query("with_keywords") keywordId: Int, @Query("sort_by") sortBy: String?
    ): Call<TVResponse?>?

    @GET("discover/tv")
    fun filterTvYear(
        @Query("api_key") apiKey: String?, @Query("language") language: String?,
        @Query("page") page: Int, @Query("limit") limit: Int,
        @Query("year") year: Int, @Query("sort_by") sortBy: String?
    ): Call<TVResponse?>?

    @GET("discover/tv")
    fun filterTvRegion(
        @Query("api_key") apiKey: String?, @Query("language") language: String?,
        @Query("page") page: Int, @Query("limit") limit: Int,
        @Query("region") region: String?, @Query("sort_by") sortBy: String?
    ): Call<TVResponse?>?

    @GET("discover/tv")
    fun filterTvGenreAndYear(
        @Query("api_key") apiKey: String?, @Query("language") language: String?,
        @Query("page") page: Int, @Query("limit") limit: Int,
        @Query("with_genres") genreId: Int, @Query("year") year: Int,
        @Query("sort_by") sortBy: String?
    ): Call<TVResponse?>?

    @GET("discover/tv")
    fun filterTvGenreAndRegion(
        @Query("api_key") apiKey: String?, @Query("language") language: String?,
        @Query("page") page: Int, @Query("limit") limit: Int,
        @Query("with_genres") genreId: Int, @Query("region") region: String?,
        @Query("sort_by") sortBy: String?
    ): Call<TVResponse?>?

    @GET("discover/tv")
    fun filterTvYearAndRegion(
        @Query("api_key") apiKey: String?, @Query("language") language: String?,
        @Query("page") page: Int, @Query("limit") limit: Int,
        @Query("year") year: Int, @Query("region") region: String?,
        @Query("sort_by") sortBy: String?
    ): Call<TVResponse?>?

    @GET("trending/{media_type}/{time_window}")
    fun getTrendingPerson(
        @Path("media_type") mediaType: String?,
        @Path("time_window") timeWindow: String?,
        @Query("api_key") apiKey: String?
    ): Call<PersonResponse?>?

    @GET("tv/{series_id}/recommendations")
    fun getTVRecommendations(
        @Path("series_id") id: Int,
        @Query("api_key") apiKey: String?
    ): Call<TVResponse?>?

    @GET("tv/{series_id}/similar")
    fun getTVSimilar(
        @Path("series_id") id: Int,
        @Query("api_key") apiKey: String?
    ): Call<TVResponse?>?
}